### ISSUE TITLE
Add support for REST hook

### ISSUE THREAD

**Issue by cverna** [fedora-infra/fpdc#31]
I think it would be nice for fpdc to support webhooks, so other services could register themselft to fpdc and receive a request when a event happens (for example a release goes EOL)

This is heavily inspired by http://resthooks.org/docs/

I think having a CURD endpoint like `/api/v1/hooks` with : 

- GET : list subscriptions
- POST : create a subscription

And `api/v1/hooks/1`

- GET : get a subscription
- PUT : update a subscription
- DELETE delete a subscription

We would need to store the following things : 

### Hooks

Attribute name | Type | Description
-- | -- | --
event_name | string | name of the event subscribe too
hook_name | string | name of the service subscribing
target_url | string | url use to send the request when a event triggers
active | boolean | Use to set the hook as active (true) or inactive (false)

### Events

Attribute name | Type | Description
-- | -- | --
event_name | string | name of the event subscribe too
payload_template | string | representation from the existing API

## Use case

For example we could have bodhi register itself to listen for the releases events. For example releases events could be : 

Event name | Description
-- | -- 
release.new | A new release was added
release.status | A release status as changed (from development to ga, or go to eol)

**Comment by cverna** [comment]
## Note 
This might add some complexity to fpdc, mostly with the need to add an asynchronous queue mechanism to send payload for each events. 
So we should consider if this feature is worth adding this complexity

**Comment by cverna** [comment]
@pypingou @abompard @bowlofeggs @ralphbean @mohanboddu I would appreciate some feedback on this proposal.

**Comment by subpop** [comment]
> And `api/v1/hooks&id=1`

Minor nitpick. Unless these URLs are added manually to the router, any use of `django-rest-framework`'s generic ViewSets is going to create this as a detail view with a URL of `api/v1/hooks/1` (in other words, the ID is in the path component rather than a query parameter).

**Comment by cverna** [comment]
> > And `api/v1/hooks&id=1`
> 
> Minor nitpick. Unless these URLs are added manually to the router, any use of `django-rest-framework`'s generic ViewSets is going to create this as a detail view with a URL of `api/v1/hooks/1` (in other words, the ID is in the path component rather than a query parameter).

Updated thanks, feel free to share your feedback on this feature too :smiley:

**Comment by mohanboddu** [comment]
I love the idea and and especially bodhi can do the necessary things when it listens to certain messages.

> ## Note
> This might add some complexity to fpdc, mostly with the need to add an asynchronous queue mechanism to send payload for each events.
> So we should consider if this feature is worth adding this complexity

But this concerns me as well.

**Comment by subpop** [comment]
As a counter-argument, what does this proposal bring that cannot be served by adding message broker publishing support for the various events (`fedora-messaging`/`fedmsg`)? If `fpdc` publishes `release.new`, bodhi can listen for that event and take action accordingly. It's the same inter-application communication flow, but without the tight coupling of app-to-app.

**Comment by cverna** [comment]
> As a counter-argument, what does this proposal bring that cannot be served by adding message broker publishing support for the various events (`fedora-messaging`/`fedmsg`)? If `fpdc` publishes `release.new`, bodhi can listen for that event and take action accordingly. It's the same inter-application communication flow, but without the tight coupling of app-to-app.

Yes doing it via https://github.com/fedora-infra/fedora-messaging is an option too, I personally like the ease of which you can create a webhook service with flask for example, but using a message bus would bring other advantages

**Comment by abompard** [comment]
Yeah I was about to suggest that this may be better suited for our message bus. Maybe we can add webhooks one day but it'll cause new problems, like if you have tens of subscriptions you'll want to run the hook from an async worker, which means a queueing system, and quickly you end up with an internal messaging system.

**Comment by subpop** [comment]
> which means a queueing system, and quickly you end up with an internal messaging system.

Which more than likely triggers another internal messaging system when the worker calls the remote system's web hook endpoint.

**Comment by cverna** [comment]
Yes I agree that the added complexity of the REST hooks makes it not really worth it. Let's looks at using fedora-messaging to accomplish the same thing. Since I think we want other application to be able to receive events from fpdc instead of having to poll the APIs.

**Comment by subpop** [comment]
My initial thought would be to hook the messaging system into the save and deletion hooks for Generic ViewSets:

```
    perform_create(self, serializer) - Called by CreateModelMixin when saving a new object instance.
    perform_update(self, serializer) - Called by UpdateModelMixin when saving an existing object instance.
    perform_destroy(self, instance) - Called by DestroyModelMixin when deleting an object instance.
```

**Comment by cverna** [comment]
Cool indeed that looks like a low effort solution at first until we identify or we have request for specific events

**Comment by subpop** [comment]
I was doing some research into this a bit more, and discovered Django has a signalling concept. This would probably be better than overriding methods on the views.

https://docs.djangoproject.com/en/2.1/topics/signals/

From skimming the documentation, they sound a lot like GObject signals.
